librz/core/cmd.c

```diff
@@ -4780,20 +4780,86 @@ DEFINE_HANDLE_TS_FCN_AND_SYMBOL(redirect_command) {
 	return res;
 }
 
+static bool do_help_manual(RzCmd *cmd, const char* cmdname) {
+	char *cmdname_help = NULL;
+	char *help_msg = NULL;
+	RzCmdParsedArgs *help_pra = NULL;
+	if (cmdname) {
+		cmdname_help = rz_str_newf("%s?", cmdname);
+		if (!cmdname_help) {
+			goto help_rec_err;
+		}
+		help_pra = rz_cmd_parsed_args_newcmd(cmdname_help);
+		if (!help_pra) {
+			goto help_rec_err;
+		}
+		help_msg = rz_cmd_get_help(cmd, help_pra, true);
+		if (!help_msg) {
+			goto help_rec_err;
+		}
+		rz_cons_printf("%s", help_msg);
+
+help_rec_err:
+		free (help_msg);
+		rz_cmd_parsed_args_free(help_pra);
+		free (cmdname_help);
+	}
+	return true;
+}
+
+static bool help_search_cmd_desc_entry(RzCmd *cmd, RzCmdDesc* cd, const char *name, void *user) {
+	rz_return_val_if_fail(cd, false);
+	if (cd->type != RZ_CMD_DESC_TYPE_ARGV &&
+		cd->type != RZ_CMD_DESC_TYPE_ARGV_MODES) {
+		return true;
+	}
+	PJ* pj = (PJ*) user;
+	if (pj) {
+		rz_cmd_get_help_json(cmd, cd, name, pj);
+	} else if (cd->type == RZ_CMD_DESC_TYPE_ARGV_MODES) {
+		do_help_manual(cmd, cd->name);
+	}
+	return true;
+}
+
+RZ_IPI RzCmdStatus rz_cmd_help_search_handler(RzCore *core, int argc, const char **argv, RzOutputMode mode) {
+	bool is_interactive = rz_cons_is_interactive(); 
+	rz_cons_set_interactive(false); 
+	PJ* pj = NULL;
+	RzCmdStatus status = RZ_CMD_STATUS_OK;
+
+	RzCmdDesc *begin = argc == 2 ? rz_cmd_get_desc(core->rcmd, argv[1]) : NULL;
+	if (argc == 2 && !begin) {
+		return RZ_CMD_STATUS_WRONG_ARGS;
+	}
+
+	if (mode & RZ_OUTPUT_MODE_JSON) {
+		pj = pj_new();
+		if (!pj) {
+			status = RZ_CMD_STATUS_ERROR;
+			goto exit_status;
+		}
+		pj_o(pj);
+	}
+
+	rz_cmd_foreach_cmdname(core->rcmd, begin, help_search_cmd_desc_entry, pj);
+
+	if (mode & RZ_OUTPUT_MODE_JSON) {
+		pj_end(pj);
+		rz_cons_printf("%s\n", pj_string(pj));
+		pj_free(pj);
+	}
+exit_status:
+	rz_cons_set_interactive(is_interactive); 
+	return status;
+}
+
 DEFINE_HANDLE_TS_FCN_AND_SYMBOL(help_command) {
 	size_t node_str_len = strlen(node_string);
 	if (node_str_len >= 2 && !strcmp(node_string + node_str_len - 2, "?*")) {
-		// TODO: get recursive help from RzCmdDesc
-		int detail = 0;
-		if (node_str_len > 3 && node_string[node_str_len - 3] == '?') {
-			detail++;
-			if (node_str_len > 4 && node_string[node_str_len - 4] == '?') {
-				detail++;
-			}
-		}
-		node_string[node_str_len - 2 - detail] = '\0';
-		recursive_help(state->core, detail, node_string);
-		return RZ_CMD_STATUS_OK;
+		// this needs to be properly handled.
```

Comment:
what does it mean?

Reply:
`?` should be refactored to be a newshell command (and probably moved under a letter instead of using the question mark).

Reply:
this has nothing to do with the `?`. This piece of code is about handling things like `z?*`, but currently it does not seem to work correctly because it prints all commands and not just the `z` sub-commands.

Reply:
nono. that piece of code is the one that is handled by "?" OLDCMD

Reply:
No :/ you can see from the if condition that it does `node_str_len >= 2 && !strcmp(node_string + node_str_len - 2, "?*")) {`. Also, you can see from the old code that it was calling `recursive_help`. So I'm quite sure this code is about `<cmd>?*`.

Reply:
no. it's the code that is for eval numbers.